sw/device/silicon_creator/lib/ownership/ownership.c

```diff
@@ -0,0 +1,102 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/silicon_creator/lib/ownership/ownership.h"
+
+#include "sw/device/lib/base/hardened.h"
+#include "sw/device/lib/base/macros.h"
+#include "sw/device/lib/base/memory.h"
+#include "sw/device/silicon_creator/lib/drivers/flash_ctrl.h"
+#include "sw/device/silicon_creator/lib/error.h"
+#include "sw/device/silicon_creator/lib/ownership/ecdsa.h"
+#include "sw/device/silicon_creator/lib/ownership/owner_keys.h"
+#include "sw/device/silicon_creator/lib/ownership/ownership.h"
+
+// RAM copy of the owner INFO pages from flash.
+owner_block_t owner_page[2];
+
+rom_error_t ownership_init(void) {
+  flash_ctrl_perms_t perm = {
+      .read = kMultiBitBool4True,
+      .write = kMultiBitBool4True,
+      .erase = kMultiBitBool4True,
+  };
+  flash_ctrl_info_perms_set(&kFlashCtrlInfoPageOwnerSlot0, perm);
+  flash_ctrl_info_perms_set(&kFlashCtrlInfoPageOwnerSlot1, perm);
+  HARDENED_RETURN_IF_ERROR(flash_ctrl_info_read(
+      &kFlashCtrlInfoPageOwnerSlot0, 0,
+      sizeof(owner_page[0]) / sizeof(uint32_t), &owner_page[0]));
+  HARDENED_RETURN_IF_ERROR(flash_ctrl_info_read(
+      &kFlashCtrlInfoPageOwnerSlot1, 0,
+      sizeof(owner_page[1]) / sizeof(uint32_t), &owner_page[1]));
+
+  // TODO(cfrantz): validate owner pages.
+  // For now, just validate the signature on the page
+  size_t len = (uintptr_t)&owner_page[0].signature - (uintptr_t)&owner_page[0];
+  hardened_bool_t result;
+  result =
+      ownership_key_validate(/*page=*/0, kOwnershipKeyOwner,
+                             &owner_page[0].signature, &owner_page[0], len);
+  if (result == kHardenedBoolFalse) {
```

Comment:
should these be hardened checks?

Reply:
I'm not sure yet - I think we should delay that decision and do a hardening analysis over the whole ROM_EXT.